### PR TITLE
shutdown vttablet gracefully if tablet record disappears

### LINKED ISSUE #7261: delete master Tablet lead to Tablet crash
image

```
I0104 14:49:15.547715 151 shard_watcher.go:56] Stopping shard watch...
I0104 14:49:15.547810 151 shard_watcher.go:62] Shard watch stopped.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xbcf711]
goroutine 323 [running]:
vitess.io/vitess/go/vt/topo.(*Server).GetTablet(0xc0002163c0, 0x2778420, 0xc002e276e0, 0x0, 0x0, 0x0, 0x0)
/vt/src/vitess.io/vitess/go/vt/topo/tablet.go:227 +0x51
vitess.io/vitess/go/vt/vttablet/tabletmanager.(*TabletManager).setMasterLocked(0xc0003b2750, 0x2778420, 0xc002e276e0, 0x0, 0x0, 0x0, 0x0, 0xc000e08301, 0x271e760, 0xc00028eaf0)
/vt/src/vitess.io/vitess/go/vt/vttablet/tabletmanager/rpc_replication.go:564 +0x200
vitess.io/vitess/go/vt/vttablet/tabletmanager.(*TabletManager).SetMaster(0xc0003b2750, 0x2778420, 0xc002e276e0, 0x0, 0x0, 0x0, 0x0, 0xc0013e0101, 0x0, 0x0)
/vt/src/vitess.io/vitess/go/vt/vttablet/tabletmanager/rpc_replication.go:482 +0x11d
vitess.io/vitess/go/vt/vttablet/tabletmanager.(*TabletManager).abortMasterTerm(0xc0003b2750, 0x27783a0, 0xc000227480, 0x0, 0x0, 0x0)
/vt/src/vitess.io/vitess/go/vt/vttablet/tabletmanager/shard_sync.go:228 +0x65b
vitess.io/vitess/go/vt/vttablet/tabletmanager.(*TabletManager).shardSyncLoop(0xc0003b2750, 0x27783a0, 0xc000227480, 0xc0003f27e0, 0xc0003f28a0)
/vt/src/vitess.io/vitess/go/vt/vttablet/tabletmanager/shard_sync.go:117 +0x632
created by vitess.io/vitess/go/vt/vttablet/tabletmanager.(*TabletManager).startShardSync
/vt/src/vitess.io/vitess/go/vt/vttablet/tabletmanager/shard_sync.go:247 +0x173
```

### PR DESCRIPTION
**Pull request by deepthi**
## Description
VTTablet should exit if the tablet record has been deleted.
This PR does that by checking for the `topo.NoNode` error.
In the process, I have deleted the `proc` package which we no longer need and simplified `servenv.Run`.

I have added an endtoend test that reproduces the panic, however the test will pass even with the panic in the log. I had to manually validate the logs to ensure we don't panic any more.

EDIT: Added a unit test that fails without the fix.

## Related Issue(s)
- Fixes #7261 

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [ ] Documentation was added or is not required

## Impacted Areas in Vitess
Components that this PR will affect:

- [ ]  Query Serving
- [ ]  VReplication
- [x]  Cluster Management
- [ ]  Build/CI
- [ ]  VTAdmin

### CODE REVIEW

File: go/vt/servenv/run.go [hunk 1]
```diff
@@ -40,13 +45,16 @@ func Run(port int) {
...
-	l, err := proc.Listen(fmt.Sprintf("%v", port))
+	l, err := net.Listen("tcp", fmt.Sprintf(":%v", port))
 	if err != nil {
 		log.Exit(err)
 	}
 	go http.Serve(l, nil)
 
-	proc.Wait()
+	ExitChan := make(chan os.Signal, 1)
+	signal.Notify(ExitChan, syscall.SIGTERM, syscall.SIGUSR1, syscall.SIGINT)
```
Comment (deepthi) [inline go/vt/servenv/run.go:55]: Not sure why we need to handle `SIGUSR1`. I see no references to this anywhere, but I carried it forward from the `proc` package.

File: go/vt/vttablet/tabletmanager/tm_state.go [hunk 2]
```diff
@@ -316,6 +319,11 @@ func (ts *tmState) publishStateLocked(ctx context.Context) {
 		return nil
 	})
 	if err != nil {
+		if topo.IsErrType(err, topo.NoNode) { // someone deleted the tablet record under us. Gracefully exit
+			log.Error("Tablet record has disappeared, shutting down")
+			servenv.ExitChan <- syscall.SIGTERM
```
Comment (deepthi) [inline go/vt/vttablet/tabletmanager/tm_state.go:324]: Perhaps we should use `SIGUSR1` instead of `SIGTERM`?

Comment (shlomi-noach) [reply go/vt/vttablet/tabletmanager/tm_state.go:324]: from https://www.gnu.org/software/libc/manual/html_node/Miscellaneous-Signals.html:

> The SIGUSR1 and SIGUSR2 signals are set aside for you to use any way you want. They’re useful for simple interprocess communication, if you write a signal handler for them in the program that receives the signal.

Which means `SIGUSR1` can mean _anything_, whereas `SIGTERM` has a clear expected behavior, which is to gracefully shotdown. For this reason. let's keep `SIGTERM` as it provides better clarity.

**Review by shlomi-noach — APPROVED** [review]
:boat:

File: go/vt/servenv/run.go (hunk 1 above)
Comment (sougou) [reply go/vt/servenv/run.go:55]: `SIGUSR1` was used as a custom signal by proc. So, we can drop it now.

File: go/vt/vttablet/tabletmanager/tm_state.go (hunk 2 above)
Comment (sougou) [reply go/vt/vttablet/tabletmanager/tm_state.go:324]: SGTM.

**Review by sougou — APPROVED** [review]

File: go/vt/servenv/run.go (hunk 1 above)
Comment (deepthi) [reply go/vt/servenv/run.go]: I've deleted it.

**Comment by deepthi** [comment]
Feedback has been addressed and all tests are passing. Merging now.

**Comment by yangxuanjia** [comment]
hi @sougou,
I maybe found a bug at  https://github.com/vitessio/vitess/pull/7563
At go/vt/servenv/run.go,  it define a global var as ExitChan chan os.Signal
But in func Run(port int) {  ExitChan := make(chan os.Signal, 1)  } , ExitChan  is redefined again.
So in other place,  they pass a signal to servenv.ExitChan <- XX,  which will not make vttablet grace shutdown.
Or it's has other reason?

**Pull request merged**